mlir/lib/Conversion/EmulateFp8ExtTrunc/EmulateFp8ExtTrunc.cpp

```diff
@@ -365,14 +384,264 @@ static FlatSymbolRefAttr makeFp8TruncFunction(Location loc, FloatType outType,
+  SmallVector<int32_t> caseLabels({0, static_cast<int>(2147483648)});
+  SmallVector<Block *> caseSuccessors({ret, bb5});
+  SmallVector<ValueRange> caseOperands({ValueRange{i8Const(0)}, ValueRange{}});
+  b.create<cf::SwitchOp>(loc, bits, bb6, ValueRange{}, caseLabels,
+                         caseSuccessors, caseOperands);
+
+  b.setInsertionPointToStart(bb5);
+  b.create<cf::BranchOp>(ret, i8Const(-128));
+
+  Block *bb7 = func.addBlock();
+  Block *bb8 = func.addBlock();
+  b.setInsertionPointToStart(bb6);
+  Value add41 = b.create<AddIOp>(eWidth, i32Const(-1));
+  Value shl42 = b.create<ShLIOp>(i32Const(-1), add41);
+  Value cmp43 = b.create<CmpIOp>(CmpIPredicate::eq, and24, i32Const(0));
+  Value cmp44 = b.create<CmpIOp>(CmpIPredicate::ne, and22, i32Const(0));
+  Value and45 = b.create<AndIOp>(cmp44, cmp43);
+  b.create<cf::CondBranchOp>(and45, bb7, bb8);
+
+  Block *bb9 = func.addBlock();
+  bb9->addArgument(i32, loc);
+  bb9->addArgument(i32, loc);
+  bb9->addArgument(i32, loc);
+  b.setInsertionPointToStart(bb7);
+  Value add46 = b.create<AddIOp>(shl42, i32Const(128));
+  b.create<cf::BranchOp>(bb9, ValueRange{i32Const(-126), add46, and22});
+
+  b.setInsertionPointToStart(bb8);
+  Value add47 = b.create<AddIOp>(shl42, i32Const(2));
+  Value add48 = b.create<AddIOp>(and24, i32Const(-127));
+  Value cmp49 = b.create<CmpIOp>(CmpIPredicate::sgt, add48, add47);
+  Value sub50 = b.create<SubIOp>(add47, add48);
+  Value or51  = b.create<OrIOp>(and22, i32Const(0x800000));
+  Value select52 = b.create<SelectOp>(cmp49, i32Const(0), sub50);
+  b.create<cf::BranchOp>(bb9, ValueRange{add48, select52, or51});
+
+  Block *bb10 = func.addBlock();
+  Block *bb11 = func.addBlock();
+  Block *bb12 = func.addBlock();
+  bb12->addArgument(i32, loc);
+  b.setInsertionPointToStart(bb9);
+  Value bb9arg0 = bb9->getArgument(0);
+  Value bb9arg1 = bb9->getArgument(1);
+  Value bb9arg2 = bb9->getArgument(2);
+  Value sub56 = b.create<SubIOp>(i32Const(23), mWidth);
+  Value add57 = b.create<AddIOp>(bb9arg1, sub56);
+  Value min58 = b.create<MinUIOp>(add57, i32Const(31));
+  Value shl59 = b.create<ShLIOp>(i32Const(-1), min58);
+  Value xor60 = b.create<XOrIOp>(shl59, i32Const(-1));
+  Value and61 = b.create<AndIOp>(bb9arg2, xor60);
+  Value add62 = b.create<AddIOp>(add57, i32Const(-1));
+  Value min63 = b.create<MinUIOp>(add62, i32Const(31));
+  Value shl64 = b.create<ShLIOp>(i32Const(1), min63);
+  Value cmp65 = b.create<CmpIOp>(CmpIPredicate::eq, and61, shl64);
+  Value cmp66 = b.create<CmpIOp>(CmpIPredicate::sgt, bb9arg1, i32Const(0));
+  b.create<cf::CondBranchOp>(cmp66, bb10, bb11);
+
+  b.setInsertionPointToStart(bb10);
+  Value min67 = b.create<MinUIOp>(bb9arg1, i32Const(31));
+  Value shr68 = b.create<ShRUIOp>(bb9arg2, min67);
+  b.create<cf::BranchOp>(bb12, ValueRange{shr68});
+
+  b.setInsertionPointToStart(bb11);
+  Value cmp69 = b.create<CmpIOp>(CmpIPredicate::eq, bb9arg1, i32Const(-1));
+  Value zext70 = b.create<ExtUIOp>(i32, cmp69);
```

Comment:
I believe that ExtUIOp turns into the llvm.zext I saw in the clang output (and similarly with sext and ExtSIOp) but I didn't trace the entire progression to confirm it.

Reply:
Yep, that's the one